### PR TITLE
Add support for preserving text attributes on enter

### LINKED ISSUE #40: Problems with text attributes in newly created lines
mar-10-2017 13-41-57

You can see a couple of things which got wrong here:

1. The style of the first line was not copied to the next line.
2. The style applied when a line was empty (so it was applied to the block) was copied – which means that we're copying block element styles (which is fine, in general) but it also means that once we started typing in that empty block the style was not moved to the text (which isn't fine).
3. The problem mentioned in 2. explains why after removing bold from the text and pressing enter, there's bold in the last block – it's taken from the paragraph itself (which shouldn't have the style at that point).

So, things to fix:

1. The style should be removed from a block once something is inserted in it. Who's job should it be? Can we actually understand which attrs applied to a block are inline attrs? I remeber we discussed this but I don't know how we implemented it.
2. Enter feature should copy style of the selection when moving it to the new block.

---

👍 if you want this issue to be fixed

### LINKED ISSUE #4562: Problems with text attributes in newly created lines
mar-10-2017 13-41-57

You can see a couple of things which got wrong here:

1. The style of the first line was not copied to the next line.
2. The style applied when a line was empty (so it was applied to the block) was copied – which means that we're copying block element styles (which is fine, in general) but it also means that once we started typing in that empty block the style was not moved to the text (which isn't fine).
3. The problem mentioned in 2. explains why after removing bold from the text and pressing enter, there's bold in the last block – it's taken from the paragraph itself (which shouldn't have the style at that point).

So, things to fix:

1. The style should be removed from a block once something is inserted in it. Who's job should it be? Can we actually understand which attrs applied to a block are inline attrs? I remeber we discussed this but I don't know how we implemented it.
2. Enter feature should copy style of the selection when moving it to the new block.

---

👍 if you want this issue to be fixed

### PR DESCRIPTION
**Pull request by msamsel**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Add support for respecting `copyOnEnter` attribute's parameter. Closes ckeditor/ckeditor5#4562.

---

### Additional information
Dependent PR:
- ~ckeditor/ckeditor5-alignment#41~
- ckeditor/ckeditor5-basic-styles#92
- ckeditor/ckeditor5-font#48
- ckeditor/ckeditor5-engine#1741

### CODE REVIEW

File: tests/entercommand.js [hunk 1]
```diff
@@ -87,6 +87,31 @@ describe( 'EnterCommand', () => {
 				'<p>x</p><p>[]</p><p>y</p>',
 				'<p>x</p><p></p><p>[]</p><p>y</p>'
 			);
+
+			describe( 'copyOnEnter', () => {
+				beforeEach( () => {
+					schema.extend( '$text', { allowAttributes: 'foo' } );
```
Comment (msamsel) [inline tests/entercommand.js]: When I look back at my code, I start to see some issues :D.
Here probably `bar` attribute should be allowed, but defined without `copyOnEnter` property.

File: tests/entercommand.js [hunk 2]
```diff
@@ -87,6 +87,31 @@ describe( 'EnterCommand', () => {
...
+				} );
+
+				test(
+					'allowed attributes are copied',
+					'<p><$text foo="true">test[]</$text></p>',
+					'<p><$text foo="true">test</$text></p><p selection:foo="true"><$text foo="true">[]</$text></p>'
+				);
+
+				test(
+					'unknown attributes are disabled',
```
Comment (jodator) [inline tests/entercommand.js]:
```suggestion
					'unknown attributes are not copied',
```

File: tests/entercommand.js (hunk 1 above)
Comment (jodator) [inline tests/entercommand.js]: As you wrote: this `bar` should be allowed also.

File: src/entercommand.js [hunk 3]
```diff
@@ -84,3 +87,11 @@ function splitBlock( writer, splitPos ) {
 	writer.split( splitPos );
 	writer.setSelection( splitPos.parent.nextSibling, 0 );
 }
+
+function* getAllowedAttributes( schema, allAttributes ) {
```
Comment (jodator) [inline src/entercommand.js]: The name is slightly misleading IMO. The name suggest `schema.getAllowedAttributes()` use (well getting allowed attributes in conjunction with schema). But this methods returns attributes that are copied on enter.

I'd go with `getCopyOnEnterAttributes()`.

File: src/entercommand.js [hunk 4]
```diff
@@ -56,7 +56,10 @@ function enterBlock( model, writer, selection, schema ) {
 	}
 
 	if ( isSelectionEmpty ) {
+		// List of text attributes copied to new line/block.
+		const filteredAttr = getAllowedAttributes( writer.model.schema, selection.getAttributes() );
```
Comment (jodator) [inline src/entercommand.js]: AFAIR no `Attr` but `Attributes` (maybe not so inside method as in public API).

Maybe go with  `attributesToCopy` and remove the comment? It should be pretty obious then what happens here so the comment would be not needed. (with conjuction of method rename below).

**Review by jodator — CHANGES_REQUESTED** [review]
Small comments to the code. I'll check the behavior in the meantime as well.

**Comment by progmancod** [comment]
Is it possible to set this same behavior to the heading plugin? I mean, copy the element class.

**Comment by Reinmar** [comment]
> Is it possible to set this same behavior to the heading plugin? I mean, copy the element class.

That's a separate piece (subfeature). Here we're copying text attributes. Copying element attributes would be a separate ticket.

**Comment by msamsel** [comment]
@Mgsy here @jodator already start a review, however, you can also take a look on functionality part. Maybe there exist some not obvious relations and connections, which you might notice.

**Comment by Mgsy** [comment]
## Steps to reproduce
1. Set the following data: `editor.setData("<p>test</p>")`.
1. Put the caret at the end of "test".
1. Activate bold.
1. Press <kbd>Shift + Enter</kbd>.

## Current result
The attribute is lost.

## GIF
bug_cke5

File: src/shiftentercommand.js [hunk 5]
```diff
@@ -81,7 +82,11 @@ function softBreakAction( model, writer, selection ) {
 	const isContainedWithinOneElement = ( startElement == endElement );
 
 	if ( isSelectionEmpty ) {
+		const attributesToCopy = getCopyOnEnterAttributes( model.schema, selection.getAttributes() );
 		insertBreak( writer, range.end );
+		// transforms: [ [ key1, value1 ], ... ] => [ key1, ... ]
```
Comment (jodator) [inline src/shiftentercommand.js]:
```js
selection.getAttributeKeys()
```
:stuck_out_tongue_winking_eye: 

ps.: the comment is also reduntant IMO - also wrongly written (should start with capital letter and have dot at the end.

File: src/utils.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module enter/utils
+ */
+
+/**
+ * Returns iterator which has filtered attributes read from {@link module:engine/model/selection~Selection#getAttributes selection}.
```
Comment (jodator) [inline src/utils.js]: Should be: `Returns iterable...` as other docs for similar constructs.

Also might just skip `iterable blah blah` and just describe that "Returns attributes that should be preserved on the enter key. Filtering..."

File: src/utils.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * Filtering is realized based on `copyOnEnter` attribute property. Read more about attribute properties
+ * {@link module:engine/model/schema~Schema#setAttributeProperties here}.
+ *
+ * @param {module:engine/model/schema~Schema} schema
+ * @param {Iterable.<*>} allAttributes iterator with attributes of current selection.
```
Comment (jodator) [inline src/utils.js]:
```suggestion
 * @param {Iterable.<*>} allAttributes Attributes to filter.
```

File: src/utils.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ * @returns {Iterable.<*>} filtered attributes which has `copyOnEnter` property.
```
Comment (jodator) [inline src/utils.js]:
```suggestion
 * @returns {Iterable.<*>}
```
AFAICS we don't copy the method description. Most methods do not have a similar description.

File: tests/utils.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import ModelTestEditor from '@ckeditor/ckeditor5-core/tests/_utils/modeltesteditor';
+
+describe( 'utils', () => {
+	describe( 'getCopyOnEnterAttributes()', () => {
+		it( 'filters attributes with copyOnEnter property', () => {
+			return ModelTestEditor.create()
+				.then( editor => {
+					const schema = editor.model.schema;
+
+					schema.register( 'foo', { inheritAllFrom: '$block' } );
```
Comment (jodator) [inline tests/utils.js]: This line registers a schema object (not an attribute) which is a `block`, similar to `paragraph`.

AFAIR only the `schema.setAttributeProperties()` is required but to have this more elegant you could add:

```js
model.schema.extend( '$text', P allowAttributes: [ 'foo', 'bar', 'baz' ] } );
```

as this behavior is for text attributes only.

**Review by jodator — CHANGES_REQUESTED** [review]
Small updates to the tests, docs and one API usage.

File: src/shiftentercommand.js (hunk 5 above)
Comment (jodator) [reply src/shiftentercommand.js]: Also the code :P - fixed.

**Review by jodator — APPROVED** [review]
All is good now. You only missed the `selection.getAttributeKeys()` use instead of custom keys retrieval from `selection.getAttributes()`.

**Comment by msamsel** [comment]
Yeah, I missed this method somehow 🤦‍♂.

**Pull request merged**